### DISCUSSION TITLE
what z_upgrade  thread do

### DISCUSSION THREAD

**Discussion by yuanfangcc** [openzfs/zfs#15176] [Q&A]
The z_upgrade thread traverses all dnodes when updating objset features. What does it do when it traverses  these dnodes?
The CMD zpool iostat has read and write traffic when z_upgrade thread work.

Does it  read all the Dnodes from disk and rewrite them to disk?
What it  read and write?

I don't understand this part of the code. Can you help me to answer it? Thank you!

Function:  `dmu_objset_space_upgrade`

**Comment by rincebrain** [comment]
From memory, it simply dirties them, and being dirty means it recalculates the quota metadata.

**Reply by yuanfangcc** [reply]
Did it read the dnode from disk and then recalculates the quota metadata before writing it to disk?

**Reply by rincebrain** [reply]
If it's not already cached in memory, yes, from (my) memory.

**Reply by yuanfangcc** [reply]
Thanks for your answer.
There are still some questions, is it the entire dnode that gets dirty, or just the bonus of the dnode?
Which data structure storage the quota metadata?

**Reply by rincebrain** [reply]
again, from memory, so someone might roll up and say i'm wrong

whole dnode gets dirtied

there's a ZAP that tracks the quota metadata for the entire dataset, if i recall, per quota type

**Reply by yuanfangcc** [reply]
Fine,Thank you for your patience.
